src/webview/html/messageAsHtml.js

```diff
@@ -72,13 +77,87 @@ $!${messageReactionListAsHtml(reactions, ownUser.user_id, allImageEmojiById)}
 `;
 };
 
-const widgetBody = (message: Message | Outbox) => template`
+/**
+ * Render the body of a message that has submessages.
+ *
+ * Must not be called on a message without any submessages.
+ */
+const widgetBody = (message: Message, ownUserId: UserId) => {
+  invariant(
+    message.submessages !== undefined && message.submessages.length > 0,
+    'should have submessages',
+  );
+
+  const widgets = message.submessages
+    .filter(submessage => submessage.msg_type === 'widget')
+    .map(submessage => ({
+      ...submessage,
+      content: JSON.parse(submessage.content),
+    }));
+
+  const poll_widgets = widgets.filter(widget => widget.content.widget_type === 'poll');
+  // "events" are submessages other than the main widget submessage - for
+  // instance, `vote`, `new_option`, and similar submessages. These are not
+  // documented, look at the PollData code if you're interested in what the
+  // possibilities are.
+  const events = widgets.filter(widget => 'type' in widget.content);
+
+  if (poll_widgets.length !== 1) {
+    return template`
 $!${message.content}
-<div class="special-message"
- ><p>Interactive message</p
- ><p>To use, open on web or desktop</p
-></div>
+<div class="special-message">
+  <p>Interactive message</p>
+  <p>To use, open on web or desktop</p>
+</div>
 `;
+  }
+  const poll_widget = poll_widgets[0];
+
+  if (!poll_widget.content || !poll_widget.content.extra_data) {
+    // We don't expect this to happen in general, but there are some malformed
+    // messages lying around that will trigger this.
```

Comment:
It might be good for the comment (or possibly the commit message) to point to the specific message(s) we know about? Without that, I could see the check lingering far after the problem gets fixed, because no one knows whether it's been fixed and there's no clear path to finding that out.

Or possibly the fact that messages are out there in the wild is enough to say that this logic has to remain permanently, since servers don't generally re-render old messages when the rendering logic changes ([doc](https://github.com/zulip/zulip-mobile/blob/master/docs/style.md#html)). Which would be a slightly unfortunate side effect on the codebase.

Reply:
IMO it should remain as long as there are messages like this out there. (Also, clients can send messages like this, and we don't want this as a DOS vector). Unfortunate, yeah, but there's not much we can do about it :/

Reply:
Makes sense. Before merging, what do you think about seeing what Steve says, with a timeout on that of a few days?